eclass/secureboot.eclass

```diff
@@ -0,0 +1,171 @@
+# installs the bootloaders and kernels to ESP always uses a signed version.
+# This prevents Secure Boot from accidentally breaking when upgrading the
+# kernel or the bootloader.
+#
+# Example use
+# @CODE
+# src_install() {
+# 	default
+# 	secureboot_sign_efi_file in.efi out.efi.signed
+# }
+# @CODE
+#
+# Or
+# @CODE
+# src_install() {
+# 	default
+# 	secureboot_auto_sign
+# }
+# @CODE
+
+case ${EAPI} in
+	7|8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI:-0} not supported" ;;
+esac
+
+IUSE="secureboot"
+BDEPEND="secureboot? ( app-crypt/sbsigntools )"
+
+# @ECLASS_VARIABLE: SECUREBOOT_SIGN_KEY
+# @USER_VARIABLE
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Used with USE=secureboot.  Should be set to the path of the private
+# key in PEM format to use, or a PKCS#11 URI.
+#
+# @ECLASS_VARIABLE: SECUREBOOT_SIGN_CERT
+# @USER_VARIABLE
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Used with USE=secureboot.  Should be set to the path of the public
+# key certificate in PEM format to use.
+
+if [[ -z ${_SECUREBOOT_ECLASS} ]]; then
+
+# @FUNCTION: _secureboot_warn_if_unset
+# @INTERNAL
+# @DESCRIPTION:
+# Prints a warning if USE=secureboot is enabled but the required user variables
+# are unset.
+_secureboot_warn_if_unset() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	if [[ -z ${SECUREBOOT_SIGN_KEY} || -z ${SECUREBOOT_SIGN_CERT} ]]; then
+		ewarn "USE=secureboot enabled but SECUREBOOT_SIGN_KEY and/or "
+		ewarn "SECUREBOOT_SIGN_CERT not set. Not signing efi files."
+		return 1
+	else
+		return 0
+	fi
+}
+
+# @FUNCTION: secureboot_pkg_setup
+# @DESCRIPTION:
+# Checks if required user variables are set before starting the build
+secureboot_pkg_setup() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	if [[ ${MERGE_TYPE} != binary ]]; then
+		_secureboot_warn_if_unset
+	fi
+}
+
+# @FUNCTION: secureboot_sign_efi_file
+# @USAGE: <input file> <output file>
+# @DESCRIPTION:
+# Sign a file using sbsign and the requested key/certificate.
+# If the file is already signed with our key then skip.
+secureboot_sign_efi_file() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	local input_file=${1}
+	local output_file=${2}
+
+	_secureboot_warn_if_unset || return
+
+	if has_version app-crypt/sbsigntools; then
+		ebegin "Signing ${input_file}"
+		local return=1
+		if sbverify "${input_file}" --cert ${SECUREBOOT_SIGN_CERT} &> /dev/null; then
+			ewarn "${input_file} already signed, skipping"
+			return
+		else
+			local args=" --key=${SECUREBOOT_SIGN_KEY} --cert=${SECUREBOOT_SIGN_CERT} "
+			if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
+				args+=" --engine=pkcs11 "
+			else
+				[[ -f ${SECUREBOOT_SIGN_KEY} ]] || die \
+					"SECUREBOOT_SIGN_KEY=${SECUREBOOT_SIGN_KEY} not found"
+			fi
+			[[ -f ${SECUREBOOT_SIGN_CERT} ]] || die \
+					"SECUREBOOT_SIGN_CERT=${SECUREBOOT_SIGN_CERT} not found"
+
+			sbsign ${args} ${input_file} --output ${output_file}
```

Comment:
Should use an array for `"${args[@]}"` (i.e. SECUREBOOT_SIGN_KEY could potentially have spaces in the path and it'd work fine with an array), and more quoting to prevent splitting on input/output file too.